### PR TITLE
Editor/Screenshots: New - capture screenshot on publish

### PR DESCRIPTION
**Pull request by Nevoss**
## PR Checklist

- [x] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Description
An explanation of what is done in this PR

* On publish it creates a screenshot of the current view of the post

## Test instructions
This PR can be tested by following these steps:

* Open a post with the editor
* Make some changes as you wish
* Publish the post
* Make sure that screenshot is created at the upload directory
* Make sure that the screenshot look the same as the preview of the post.

## Quality assurance

- [x] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+			...{ post: { id: elementorFrontendConfig.post.id } }
+		};
+
+		if ( ! this.$elementor.length ) {
+			console.warn( 'Screenshots: Elementor content was not found.' );
+			return;
+		}
+
+		return Promise.resolve()
+			.then( this.handleIFrames.bind( this ) )
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:43]: Try remove the promise if not needed.

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+			.then( this.loadExternalCss.bind( this ) )
+			.then( this.hideUnnecessaryElements.bind( this ) )
+			.then( this.createImage.bind( this ) )
+			.then( this.createImageElement.bind( this ) )
+			.then( this.cropCanvas.bind( this ) )
+			.then( this.save.bind( this ) )
+	}
+
+	/**
+	 * The html-to-image library can not snapshot the iframes.
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:54]:
```suggestion
	 * Html-to-image libraries can not snapshot the iframes.
```

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	 * so this method convert all the iframes with something else.
+	 *
+	 * @returns {Promise<void>}
+	 */
+	handleIFrames() {
+		const $self = this;
+
+		this.$elementor.find( 'iframe' ).each( function () {
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:62]: Try use arrow function + remove the $self.

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+			$this.find( '> *' ).not( $this.find( '> :first-child' ) ).each( function () {
+				jQuery( this ).remove()
+			} );
+		} );
+
+		return Promise.resolve();
+	}
+
+	/**
+	 * css from another server cannot be loaded.
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:128]:
```suggestion
	 * CSS from another server cannot be loaded.
```

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+	 * that is why we fetch the content of the css and set it as an inline css.
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:129]:
```suggestion
	 * that is why it needed to be fetch the content of the css and set it as an inline css.
```

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+	 *
+	 * @returns {Promise<unknown[]>}
+	 */
+	loadExternalCss() {
+		const $self = this;
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:134]: Try remove.

File: modules/screenshots/assets/js/frontend/screenshot.js [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+	/**
+	 * Creates a png image
+	 *
+	 * @returns {Promise<unknown>}
+	 */
+	createImage() {
+		return new Promise( ( resolve ) => {
+			setTimeout( () => {
+				domtoimage.toPng( this.$elementor.parents( 'body' ).get( 0 ), {} )
```
Comment (matipojo) [inline modules/screenshots/assets/js/frontend/screenshot.js:169]: `document.body`.

File: modules/screenshots/module.php [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		wp_update_attachment_metadata(
+			$attachment_id,
+			wp_generate_attachment_metadata( $attachment_id, $upload['file'] )
+		);
+
+		return $upload['url'];
+	}
+
+	public function enqueue_scripts() {
+		if ( ! isset( $_REQUEST['elementor-screenshot'] ) || ! User::is_current_user_can_edit() ) { // WPCS: CSRF ok.
```
Comment (matipojo) [inline modules/screenshots/module.php:91]:
```suggestion
		if ( ! isset( $_REQUEST['elementor-screenshot'] ) || ! User::is_current_user_can_edit() ) {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/screenshots/assets/js/editor/component.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+	defaultCommandsInternal() {
+		return this.importCommands( internalCommands );
+	}
+
+	defaultHooks() {
+		return this.importHooks( hooks );
+	}
+
+	isCapturingScreenshot() {
```
Comment (matipojo) [inline modules/screenshots/assets/js/editor/component.js:37]: It's a public property, so it don't needs a getter.

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	/**
+	 * CSS from another server cannot be loaded,
+	 * that is the reason behind this method, fetching the content of the CSS and set it as an inline css.
+	 *
+	 * @returns {Promise<unknown[]>}
+	 */
+	loadExternalCss() {
+		const selector = this.config.allowedCssUrls.map( ( allowedCssUrl ) => {
```
Comment (matipojo) [inline modules/screenshots/assets/js/preview/screenshot.js:153]: What about adding a proxy URL to the original external link?

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+	 * Load one css file.
+	 *
+	 * @param url
+	 * @returns {Promise<void>}
+	 */
+	loadCss( url ) {
+		return fetch( url )
+			.then( ( response ) => response.text() )
+			.then( ( data ) => {
+				// This is a specific use case when we need
```
Comment (matipojo) [inline modules/screenshots/assets/js/preview/screenshot.js:277]: We :)

File: modules/screenshots/module.php [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+class Module extends BaseModule {
+
+	public function get_name() {
+		return 'screenshots';
+	}
+
+	public function screenshot_proxy() {
+		if ( ! wp_verify_nonce( $_GET['nonce'], 'screenshot_proxy' ) ) {
```
Comment (matipojo) [inline modules/screenshots/module.php:20]: Do we have a pattern for allowed URLs? (like `/(png|jpg|css)/`)

File: modules/screenshots/module.php [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+	public function ajax_save( $data ) {
+		if ( empty( $data['screenshot'] ) ) {
+			return false;
+		}
+
+		$post_id = $data['post_id'];
+
+		$file_content = substr( $data['screenshot'], strlen( 'data:image/png;base64,' ) );
+		$file_name = 'Elementor Post Screenshot ' . $post_id . '.png';
```
Comment (matipojo) [inline modules/screenshots/module.php:41]: I think your preview fix was good:
```suggestion
		$file_name = 'Elementor-post-screenshot-' . $post_id . '.png';
```

File: modules/screenshots/module.php [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+		$over_write_file_name_callback = function () use ( $file_name ) {
+			return $file_name;
+		};
+
+		add_filter( 'wp_unique_filename', $over_write_file_name_callback );
+
+		$upload = wp_upload_bits(
```
Comment (matipojo) [inline modules/screenshots/module.php:48]: Try move all screenshots to a new folder `/wp-content/uploads/elementor/screenshots`. (create if not exist with `wp_mkdir_p`)

File: modules/screenshots/module.php [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+		wp_update_attachment_metadata(
+			$attachment_id,
+			wp_generate_attachment_metadata( $attachment_id, $upload['file'] )
+		);
+
+		return $upload['url'];
+	}
+
+	public function enqueue_scripts() {
+		if ( ! $this->is_in_screenshot_mode() || ! User::is_current_user_can_edit() ) {
```
Comment (matipojo) [inline modules/screenshots/module.php:91]:
```suggestion
		if ( ! $this->is_screenshot_mode() || ! User::is_current_user_can_edit() ) {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+		this.$elementor = jQuery( ElementorScreenshotConfig.selector );
+		this.config = {
+			...this.config,
+			...ElementorScreenshotConfig,
+		};
+
+		if ( ! this.$elementor.length ) {
+			elementorCommon.helpers.consoleWarn( 'Screenshots: Elementor content was not found.' );
+
+			return;
```
Comment (matipojo) [inline modules/screenshots/assets/js/preview/screenshot.js:52]: Run `screenshotFailed` in order to clear the timer.

File: modules/screenshots/assets/js/editor/commands/internal/capture.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+		jQuery( 'body' ).append( iframe );
+
+		// The iframe send an event when the screenshot process complete
+		// then the command send a notice to the component about it.
+		window.onmessage = ( message ) => {
+			if ( ! message.data.name || message.data.name !== 'capture-screenshot-done' ) {
+				return;
+			}
+
+			this.component.isCapturingScreenshot = false;
```
Comment (matipojo) [inline modules/screenshots/assets/js/editor/commands/internal/capture.js:18]: Remove the iframe element.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+			crop: {
+				width: 1200,
+				height: 1500,
+			},
+			excludedCssUrls: [
+				'https://kit-pro.fontawesome.com',
+			],
+			timeout: 15000, // Wait until screenshot taken or fail in 15 secs.
+			render_timeout: 5000, // Wait until all the element will be loaded or 5 sec and then take screenshot.
+			timerLabel: null,
```
Comment (matipojo) [inline modules/screenshots/assets/js/preview/screenshot.js:25]: Keep all config keys with `snake_case` format.

File: modules/screenshots/screenshot.php [hunk 19]
```diff
@@ -0,0 +1,199 @@
+<?php
+
```
Comment (matipojo) [inline modules/screenshots/screenshot.php:2]: Remove all top empty lines.

File: modules/screenshots/screenshot.php [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+	/**
+	 * Get the file name,
+	 * if not exists will generate it.
+	 *
+	 * @return string
+	 */
+	public function get_file_name() {
+		if ( ! $this->file_name ) {
+			$now = ( new \DateTime() )->format( 'Y-m-dTH:i:s' );
```
Comment (matipojo) [inline modules/screenshots/screenshot.php:149]: Use valid file name characters.
```suggestion
			$now = ( new \DateTime() )->format( 'Y-m-d-H-i-s' );
```

File: tests/phpunit/elementor/modules/screenshots/module/test-ajax-action.php [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	public function test_should_return_false_if_there_is_no_screenshot() {
+		$response = $this->module->ajax_save( [ 'post_id' => $this->post->ID ] );
+
+		$this->assertFalse( $response );
+	}
+
+	public function test_should_return_false_if_there_is_no_post_id() {
+		$response = $this->module->ajax_save( [ 'screenshot' => 'adsasd' ] );
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/screenshots/module/test-ajax-action.php:37]:
```suggestion
		$response = $this->module->ajax_save( [ 'screenshot' => 'invalid-id' ] );
```

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — DISMISSED** [review]

**Review by matipojo — DISMISSED** [review]

File: tests/phpunit/elementor/modules/screenshots/module/test-ajax-action.php [hunk 22]
```diff
@@ -0,0 +1,64 @@
+<?php
+
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/screenshots/module/test-ajax-action.php:2]: Remove all top empty lines.

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by kobizz — CHANGES_REQUESTED** [review]

File: modules/screenshots/assets/js/editor/commands/internal/capture.js [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+		iframe.width = '1200';
+		iframe.style = 'visibility: hidden;';
+
+		return iframe;
+	}
+
+	getIframeUrl() {
+		return elementor.config.document.urls.screenshot;
+	}
+}
```
[the remark this answers is not in the document]
Comment (matipojo) [reply modules/screenshots/assets/js/editor/commands/internal/capture.js:57]: It's not needed here.

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 24]
```diff
@@ -0,0 +1,378 @@
+/* global ElementorScreenshotConfig, jQuery */
+class Screenshot {
```
[the remark this answers is not in the document]
Comment (matipojo) [reply modules/screenshots/assets/js/preview/screenshot.js:2]: It's not a part of the editor environment. it's more like a "frontend handle" bu the handlers system is not available on static render.

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 25]
```diff
@@ -0,0 +1,378 @@
...
+		$iframe.next().remove();
+	}
+
+	/**
+	 * @param $iframe
+	 * @param $iframeMask
+	 */
+	handleYouTubeIFrame( $iframe, $iframeMask ) {
+		const regex = /^.*(?:youtu.be\/|youtube(?:-nocookie)?.com\/(?:(?:watch)??(?:.*&)?vi?=|(?:embed|v|vi|user)\/))([^?&"'>]+)/;
+		const matches = regex.exec( $iframe.attr( 'src' ) );
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/screenshots/assets/js/preview/screenshot.js]: Function removed from the code (not needed any more).

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+	}
+
+	/**
+	 * Slides should show only the first slide, all the other slides will be removed.
+	 */
+	handleSlides() {
+		this.$elementor.find( '.elementor-slides' ).each( ( index, el ) => {
+			const $this = jQuery( el );
+
+			$this.find( '> *' ).not( $this.find( '> :first-child' ) ).each( ( childIndex, childEl ) => {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/screenshots/assets/js/preview/screenshot.js]: Function removed from the code (not needed any more).

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+
+	/**
+	 * @param $iframe
+	 * @param $iframeMask
+	 */
+	handleCustomEmbedImageIFrame( $iframe, $iframeMask ) {
+		const regex = /url\(\"(.*)\"/gm;
+		const url = $iframe.next().css( 'backgroundImage' );
+		const matches = regex.exec( url );
+
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/screenshots/assets/js/preview/screenshot.js]: Function removed from the code (not needed any more).

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 28]
```diff
@@ -87,86 +92,6 @@ class Screenshot {
 		document.body.prepend( this.$elementor );
 	}
 
-	/**
-	 * Html to images libraries can not snapshot IFrames
-	 * this method convert all the IFrames to some other elements.
-	 */
-	handleIFrames() {
```
Comment (matipojo) [inline modules/screenshots/assets/js/preview/screenshot.js]: We still need it for custom iframes.

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 29]
```diff
@@ -192,6 +117,21 @@ class Screenshot {
 		} );
 	}
 
+	/**
+	 * Make a proxy to images urls that has some problems with cross origin (like youtube).
+	 */
+	loadExternalImages() {
+		const selector = this.config.include_images_urls
```
Comment (matipojo) [inline modules/screenshots/assets/js/preview/screenshot.js:124]:
```suggestion
		const selector = this.config.external_images_urls
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/screenshots/assets/js/editor/commands/internal/capture.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (kobizz) [reply modules/screenshots/assets/js/editor/commands/internal/capture.js:57]: Our standard is to use jQuery for dom manipulations and queries

File: modules/screenshots/assets/js/preview/screenshot.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (kobizz) [reply modules/screenshots/assets/js/preview/screenshot.js:4]: Our standard is to handle all our views modules the same way. There is a ViewModule base available on both ends and you should use it.

File: assets/lib/html2canvas/js/html2canvas.min.js [hunk 30]
```diff
@@ -0,0 +1,20 @@
+/*!
```
Comment (kobizz) [inline assets/lib/html2canvas/js/html2canvas.min.js:1]: This library weights 120k in its minified version. I don't think it worths it.

File: modules/screenshots/assets/js/editor/commands/internal/capture.js [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+
+		this.$iframe = null;
+	}
+
+	apply() {
+		this.component.isCapturingScreenshot = true;
+
+		this.$iframe = this.createIframe();
+
+		jQuery( 'body' ).append( this.$iframe );
```
Comment (kobizz) [inline modules/screenshots/assets/js/editor/commands/internal/capture.js:15]:
```suggestion
		elementorCommon.elements.$body.append( this.$iframe );
```

File: modules/screenshots/assets/js/editor/commands/internal/capture.js [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+			};
+
+			window.addEventListener( 'message', listener );
+		} );
+	}
+
+	createIframe() {
+		const $iframe = jQuery( '<iframe></iframe>' );
+
+		$iframe.attr( 'src', this.getIframeUrl() );
```
Comment (kobizz) [inline modules/screenshots/assets/js/editor/commands/internal/capture.js:47]: You can simplify it:

```suggestion
		const $iframe = jQuery( '<iframe>', { src: this.getIframeUrl() } );
```

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 33]
```diff
@@ -0,0 +1,352 @@
...
+			$iframe.before( $iframeMask );
+			$iframe.remove();
+		} );
+	}
+
+	/**
+	 * Hide all the element except for the target element.
+	 */
+	hideUnnecessaryElements() {
+		jQuery( 'body' ).prepend(
```
Comment (kobizz) [inline modules/screenshots/assets/js/preview/screenshot.js:157]:
```suggestion
		elementorCommon.elements.$body.prepend(
```

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 34]
```diff
@@ -0,0 +1,352 @@
...
+			this.elements.$elementor
+		);
+
+		jQuery( 'body > *' ).not( this.elements.$elementor ).css( 'display', 'none' );
```
Comment (kobizz) [inline modules/screenshots/assets/js/preview/screenshot.js:161]: Consider caching `jQuery( 'body > *' ).not( this.elements.$elementor )` inside `getDefaultElements`

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 35]
```diff
@@ -0,0 +1,352 @@
...
+			.map( ( url ) => `[href^="${ url }"]` )
+			.join( ', ' );
+
+		jQuery( 'link' ).not( notSelector ).each( ( index, el ) => {
+			const $link = jQuery( el ),
+				$newLink = $link.clone();
+
+			$newLink.attr( 'href', this.getScreenshotProxyUrl( $link.attr( 'href' ) ) );
+
+			jQuery( 'head' ).append( $newLink );
```
Comment (kobizz) [inline modules/screenshots/assets/js/preview/screenshot.js:113]: cache `jQuery( 'head' )` inside `getDefaultElements`

File: modules/screenshots/assets/js/preview/screenshot.js [hunk 36]
```diff
@@ -0,0 +1,352 @@
...
+	}
+
+	/**
+	 * Remove all the sections that should not be in the screenshot.
+	 */
+	removeUnnecessaryElements() {
+		let currentHeight = 0;
+
+		this.elements.$elementor
+			.find( ' .elementor-section-wrap > .elementor-section' )
```
Comment (kobizz) [inline modules/screenshots/assets/js/preview/screenshot.js:171]: Consider caching it inside `getDefaultElements`

File: modules/screenshots/screenshot.php [hunk 37]
```diff
@@ -0,0 +1,198 @@
...
+
+		return $this;
+	}
+
+	/**
+	 * Uploads the base64 image it self.
+	 *
+	 * @return $this
+	 */
+	public function upload() {
```
Comment (kobizz) [inline modules/screenshots/screenshot.php:74]: Add a TODO to use the planned Uploads Manager here

**Review by kobizz — CHANGES_REQUESTED** [review]

File: modules/screenshots/module.php [hunk 38]
```diff
@@ -0,0 +1,250 @@
...
+			'selector' => '.elementor-' . $post_id,
+			'nonce' => wp_create_nonce( self::SCREENSHOT_PROXY_NONCE_ACTION ),
+			'home_url' => home_url(),
+			'post_id' => $post_id,
+		];
+
+		wp_add_inline_script( 'elementor-screenshot', 'var ElementorScreenshotConfig = ' . wp_json_encode( $config ) . ';' );
+
+		$css = Post_Preview::create( $post_id );
+		$css->enqueue();
```
Comment (matipojo) [inline modules/screenshots/module.php:118]: Try remove.

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — DISMISSED** [review]

**Review by kobizz — APPROVED** [review]

**Review by matipojo — APPROVED** [review]

**Review by KingYes — APPROVED** [review]

**Pull request merged**